python/cudf_polars/cudf_polars/experimental/rapidsmpf/union.py

```diff
@@ -0,0 +1,95 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-License-Identifier: Apache-2.0
+"""Union logic for the RapidsMPF streaming runtime."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Any
+
+from rapidsmpf.streaming.core.channel import Message
+from rapidsmpf.streaming.cudf.table_chunk import TableChunk
+
+from cudf_polars.dsl.ir import Union
+from cudf_polars.experimental.rapidsmpf.dispatch import (
+    generate_ir_sub_network,
+)
+from cudf_polars.experimental.rapidsmpf.nodes import define_py_node, shutdown_on_error
+from cudf_polars.experimental.rapidsmpf.utils import (
+    ChannelManager,
+    process_children,
+)
+
+if TYPE_CHECKING:
+    from rapidsmpf.streaming.core.context import Context
+
+    from cudf_polars.dsl.ir import IR
+    from cudf_polars.experimental.rapidsmpf.core import SubNetGenerator
+    from cudf_polars.experimental.rapidsmpf.utils import ChannelPair
+
+
+@define_py_node()
+async def union_node(
+    ctx: Context,
+    ir: Union,
+    ch_out: ChannelPair,
+    *chs_in: ChannelPair,
+) -> None:
+    """
+    Union node for rapidsmpf.
+
+    Parameters
+    ----------
+    ctx
+        The context.
+    ir
+        The Union IR node.
+    ch_out
+        The output ChannelPair.
+    chs_in
+        The input ChannelPairs.
+    """
+    # TODO: Use multiple streams
+    async with shutdown_on_error(ctx, *[ch.data for ch in chs_in], ch_out.data):
+        seq_num_offset = 0
+        for ch_in in chs_in:
+            num_ch_chunks = 0
+            while (msg := await ch_in.data.recv(ctx)) is not None:
+                table_chunk = TableChunk.from_message(msg)
+                num_ch_chunks += 1
+                await ch_out.data.send(
+                    ctx,
+                    Message(
+                        TableChunk.from_pylibcudf_table(
+                            table_chunk.sequence_number + seq_num_offset,
+                            table_chunk.table_view(),
+                            table_chunk.stream,
+                            exclusive_view=True,
+                        )
+                    ),
+                )
+            seq_num_offset += num_ch_chunks
```

Comment:
Two things here:

1. sequence numbers for messages in a channel are not guaranteed to be dense in `[0, ..., num_messages_in_channel)`. A valid set of sequence numbers providing ordering is `[0, 400, 10_000]`. I think today all the sequence numbers _are_ dense. But if we need that for various algorithms we should specify it.
2. You're only reading and/or updating the metadata of the `TableChunk` so you really don't want to do `from_pylibcudf_table` to send a new thing on (because the `table_view()` call may fail because the data are currently spilled). So that means that if you can't do that today, we should add facilities for updating metadata in a chunk.

Reply:
Re 1: I propose that we use dense sequence numbers in cudf-polars, unless you have a specific application in mind that is blocked by this. It will be much easier to implement and debug algorithms if the chunks have dense indices.

Re 2: It looks like I may need to add a new `TableChunk.set_sequence_number` method to rapidsmpf?